File: mobile/test/operators/test_lrn_op.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "../test_include.h"
#include "operators/lrn_op.h"

int main() {
  paddle_mobile::framework::Loader<paddle_mobile::CPU> loader;
  auto program = loader.Load(g_googlenet);
  PADDLE_MOBILE_ENFORCE(program.originProgram != nullptr,
                        "program file read fail");

  Executor4Test<paddle_mobile::CPU,
                paddle_mobile::operators::LrnOp<paddle_mobile::CPU, float>>
      executor(program, "lrn");

  // 1. input_tensors;
  vector<Tensor> input_tensors;

  Tensor input1;
  auto input1_data = CreateInput<float>(&input1, {3, 4, 2, 2}, 0, 1);
  input_tensors.push_back(input1);

  // 2. input_names
  vector<string> input_names({
      "pool2d_0.tmp_0",
  });

  // 3. output_names
  vector<string> output_names({"pool1_norm1.tmp_1"});

  // 4. out_dims;
  vector<DDim> out_ddims;
  auto out_ddim = paddle_mobile::framework::make_ddim({3, 4, 2, 2});
  out_ddims.push_back(out_ddim);

  auto output = executor.Predict<LoDTensor>(input_tensors, input_names,
                                            output_names, out_ddims);

  auto output0_data = output[0]->data<float>();

  DLOG << " LrnOp input: ";
  for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 4; j++) {
      for (int c = 0; c < 2; c++) {
        for (int d = 0; d < 2; d++) {
          DLOGF("%f ", input1_data[i * 16 + j * 4 + c * 2 + d]);
        }
        DLOGF("\n");
      }
      DLOGF("\n");
    }
    DLOGF("\n");
  }
  DLOG << " LrnOp output: ";
  for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 4; j++) {
      for (int c = 0; c < 2; c++) {
        for (int d = 0; d < 2; d++) {
          DLOGF("%f ", output0_data[i * 16 + j * 4 + c * 2 + d]);
        }
        DLOGF("\n");
      }
      DLOGF("\n");
    }
    DLOGF("\n");
  }
  DLOG << input1_data[0] << " / ((1 + 0.00002 * ( " << input1_data[0] << "^2 + "
       << input1_data[4] << "^2 + " << input1_data[8] << "^2 ))^0.75) = ";
  DLOG << output0_data[0];
  return 0;
}

File: lite/tests/unittest_py/op/test_lrn_op.py
# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
sys.path.append('../')

from auto_scan_test import AutoScanTest, IgnoreReasons
from program_config import TensorConfig, ProgramConfig, OpConfig, CxxConfig, TargetType, PrecisionType, DataLayoutType, Place
import unittest

import hypothesis
from hypothesis import given, settings, seed, example, assume

import numpy as np
from functools import partial
import hypothesis.strategies as st


class TestLrnOp(AutoScanTest):
    def __init__(self, *args, **kwargs):
        AutoScanTest.__init__(self, *args, **kwargs)
        self.enable_testing_on_place(
            TargetType.ARM,
            PrecisionType.FP32,
            DataLayoutType.NCHW,
            thread=[1, 4])
        # opencl bug will be fix in the future
        opencl_places = [
            Place(TargetType.OpenCL, PrecisionType.FP16,
                  DataLayoutType.ImageDefault), Place(
                      TargetType.OpenCL, PrecisionType.FP16,
                      DataLayoutType.ImageFolder),
            Place(TargetType.OpenCL, PrecisionType.FP32, DataLayoutType.NCHW),
            Place(TargetType.OpenCL, PrecisionType.Any,
                  DataLayoutType.ImageDefault), Place(
                      TargetType.OpenCL, PrecisionType.Any,
                      DataLayoutType.ImageFolder),
            Place(TargetType.OpenCL, PrecisionType.Any, DataLayoutType.NCHW),
            Place(TargetType.Host, PrecisionType.FP32)
        ]
        self.enable_testing_on_place(places=opencl_places)

    def is_program_valid(self,
                         program_config: ProgramConfig,
                         predictor_config: CxxConfig) -> bool:
        return True

    def sample_program_configs(self, draw):
        #simple case
        '''
        in_shape = draw(st.sampled_from([[1, 5, 2, 4]]))
        n_ = draw(st.sampled_from([5]))
        k_ = draw(st.sampled_from([1.0]))
        alpha_ = draw(st.sampled_from([1e-4]))
        beta_ = draw(st.sampled_from([0.75, 1]))
        '''
        in_num = draw(
            st.lists(
                st.integers(
                    min_value=1, max_value=4), min_size=1, max_size=1))
        in_c_h_w = draw(
            st.lists(
                st.integers(
                    min_value=1, max_value=128),
                min_size=3,
                max_size=3))
        in_shape = in_num + in_c_h_w
        n_ = draw(st.integers(min_value=1, max_value=in_c_h_w[0]))
        k_ = draw(st.floats(min_value=1.0, max_value=10.0))
        alpha_ = draw(st.floats(min_value=1.0, max_value=10.0))
        beta_ = draw(st.floats(min_value=1.0, max_value=10.0))
        assume(n_ % 2 == 1)
        lrn_op = OpConfig(
            type="lrn",
            inputs={"X": ["input_data"]},
            outputs={"Out": ["output_data"],
                     "MidOut": ["output_data_mid"]},
            attrs={
                "n": n_,
                "k": k_,
                "alpha": alpha_,
                "beta": beta_,
                "is_test": 1
            })
        program_config = ProgramConfig(
            ops=[lrn_op],
            weights={},
            inputs={"input_data": TensorConfig(shape=in_shape)},
            outputs=["output_data"])
        return program_config

    def sample_predictor_configs(self):
        return self.get_predictor_configs(), ["lrn"], (1e-5, 1e-5)

    def add_ignore_pass_case(self):
        pass

    def test(self, *args, **kwargs):
        self.run_and_statis(quant=False, max_examples=150)


if __name__ == "__main__":
    unittest.main(argv=[''])
